python-package/lightgbm/basic.py

```diff
@@ -1444,16 +1460,21 @@ def get_data(self):
         """
         if self.handle is None:
             raise Exception("Cannot get data before construct Dataset")
-        if self.data is not None and self.used_indices is not None and self.need_slice:
-            if isinstance(self.data, np.ndarray) or scipy.sparse.issparse(self.data):
-                self.data = self.data[self.used_indices, :]
-            elif isinstance(self.data, DataFrame):
-                self.data = self.data.iloc[self.used_indices].copy()
-            elif isinstance(self.data, DataTable):
-                self.data = self.data[self.used_indices, :]
+        if self.need_slice and self.used_indices is not None and self.reference is not None:
+            self.data = self.reference.data
+            if self.data is None:
+                warnings.warn("The raw data of the subset reference is released, cannot call get_data().")
             else:
-                warnings.warn("Cannot subset {} type of raw data.\n"
-                              "Returning original raw data".format(type(self.data).__name__))
+                if isinstance(self.data, np.ndarray) or scipy.sparse.issparse(self.data):
+                    self.data = self.data[self.used_indices, :]
+                elif isinstance(self.data, DataFrame):
+                    self.data = self.data.iloc[self.used_indices].copy()
+                elif isinstance(self.data, DataTable):
+                    self.data = self.data[self.used_indices, :]
+                else:
+                    warnings.warn("Cannot subset {} type of raw data.\n"
+                                  "Returning original raw data".format(type(self.data).__name__))
+                    self.data = None
```

Comment:
Warning message says that the original data will be returned, but actually `None` will be returned always.

```
import numpy as np
import lightgbm as lgb

data = np.random.random((500, 2))
train_data = lgb.Dataset(data, free_raw_data=False)
train_data.save_binary('train_data.bin')
train_data = lgb.Dataset("train_data.bin", free_raw_data=False)
type(train_data.subset([0]).construct().get_data())
>>> NoneType
```